proposals/blittable.md

```diff
@@ -8,129 +8,97 @@
+Hash(42); // Okay
+Hash("hello") // Error: Type string does not satisfy the unmanaged constraint
+```
 
-void Hash(Time t) { 
-    ...
+Type parameters with the unmanaged constraint can use all the features available to unmanaged types: pointers, fixed, etc ... 
+
+
+``` c#
+void Hash<T>(T value) where T : unmanaged
+{
+    // Okay
+    fixed (T* p = &value) { 
+        ...
+    }
 }
 ```
 
-The lack of constraints here also make it impossible to have efficient conversions between streams of data and more structured types.  This is an operation that is common in networking stacks and serialization layers:
+This constraint will also make it possible to have efficient conversions between structured data and streams of bytes. This is an operation that is common in networking stacks and serialization layers:
 
 ``` c#
-Span<byte> Convert<T>(ref T value) where T : blittable {
+Span<byte> Convert<T>(ref T value) where T : unmanaged {
     ...
 }
 ```
 
-Such routines are advantageous because they are provably safe at compile time and allocation free.  Interop authors today can not due this (even though it's at a layer where perf is critical).  Instead they need to rely on allocating routines that have expensive runtime checks to verify values are correctly blittable.
+Such routines are advantageous because they are provably safe at compile time and allocation free.  Interop authors today can not do this (even though it's at a layer where perf is critical).  Instead they need to rely on allocating routines that have expensive runtime checks to verify values are correctly unmanaged.
 
 ## Detailed design
 [design]: #detailed-design
 
-The language will introduce a new declaration modifier named `blittable` that can be applied to `struct` definitions.  
+The language will introduce a new constraint named `unmanaged`. In order to satisfy this constraint a type must be a struct and all the fields of the type must fall into one of the following categories:
 
-``` c#
-blittable struct Point 
-{
-    public int X;
-    public int Y;
-}
-```
+- Have the type `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool`, `IntPtr` or `UIntPtr`.
+- Be any `enum` type.
+- Be a pointer type.
+- Be a user defined struct that satsifies the `unmanaged` constraint.
 
-The compiler will enforce that the fields of such a struct definition fit one of the following categories:
+Compiler generated instance fields, such as those backing auto-implemented properties, must also meet these constraints. 
 
-- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
-- Any `enum` type
-- Any pointer type which points to a `blittable` type
-- Any user defined struct explicitly declared as `blittable`
+For example:
 
-Any type fitting the definition above, or any element in the list, is considered a valid `blittable` type.
+``` c# 
+// Unmanaged type
+struct Point { 
```

Comment:
is there a way to say "unmanaged struct Point" in order to get the compiler to enforce that you don't actually make something unmanaged in the future?

Reply:
I agree. It would be useful to indicate, I declare this is an unmanaged type, so please ensure I don't break the contract

Reply:
That was included in the previous design. Abandoned it for this iteration because both of the design ideas had strong negatives:

- Require `unmanaged struct` to meet `unmanaged` constraint: huge compat issue. Loads of types which are functionally unmanaged in existing code will not meet the constraint. Could be years for it to all get in sync. 
- `unmanaged struct` declaration just enforces this type is functionally unmanaged, not required for constraint. Can be a bit confusing there is a constraint and declaration of same name here. But it's not required for a declaration to be unmanaged to meet the constraint. 

If you're interested in the second though you can always do the following:

``` c# 
static class AssemblyUtil { 
  static void RequireUnmanaged<T>() where T : unmanaged { } 
  static void Go() { 
    RequireUnmanaged<Point>();
  }
}
```

Not as nice of crouse.

Reply:
I think the second is the request here.

Just allowing any type which meets the constraint to be used is nice, and will ensure we don't have to worry about getting things in sync.

The second, however, allows me to enforce that the type always meets the constraint itself. This is useful to prevent future refactorings from breaking it, which may be needed/useful for larger frameworks.

Reply:
An analyzer could also be used here. However, if we go that route, I would like a coordinated effort to ship a CoreFX `UnmanagedType` attribute and analyzer SxS with the release, so we don't end up with 3-4 different ways/analyzers for doing the check.